src/Analyzers/Core/Analyzers/SimplifyLinqExpression/SimplifyLinqExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+#if CODE_STYLE
+using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
+#endif
+
+namespace Microsoft.CodeAnalysis.SimplifyLinqExpression
+{
+    internal abstract class SimplifyLinqExpressionDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private readonly ImmutableArray<string> _nonEnumerableReturningLinqMethodNames =
+            ImmutableArray.Create(
+                nameof(Enumerable.First),
+                nameof(Enumerable.Last),
+                nameof(Enumerable.Single),
+                nameof(Enumerable.Any),
+                nameof(Enumerable.Count),
+                nameof(Enumerable.SingleOrDefault),
+                nameof(Enumerable.FirstOrDefault),
+                nameof(Enumerable.LastOrDefault)
+            );
+
+        public SimplifyLinqExpressionDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.SimplifyLinqExpressionDiagnosticId,
+                   EnforceOnBuildValues.SimplifyLinq,
+                   option: null,
+                   title: new LocalizableResourceString(
+                       nameOfLocalizableResource: nameof(AnalyzersResources.Simplify_Linq_expression),
+                       resourceManager: AnalyzersResources.ResourceManager,
+                       resourceSource: typeof(AnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(OnCompilationStart);
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if ((typeof(Enumerable)?.FullName is not string fullyQualifiedName) ||
+                (context.Compilation.GetTypeByMetadataName(fullyQualifiedName) is not INamedTypeSymbol enumerableType))
+            {
+                return;
+            }
```

Comment:
```suggestion
            var enumerableType = context.Compilation.GetTypeByMetadataName(typeof(Enumerable)?.FullName);
            if (enumerableType == null)
                return;
```

Reply:
GetTypeByMetadataName does not accept nullable strings

Reply:
`typeof(Enumerable)?.FullName` will always be non-null.  if something doesn't understand that, use `!` to shut it up :)